// @enact/sandstone/internal/SharedStateDecorator/SharedStateDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SharedStateDecorator = exports.SharedState = void 0;

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = require("react");

var _jsxRuntime = require("react/jsx-runtime");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var SharedState = /*#__PURE__*/(0, _react.createContext)(null);
exports.SharedState = SharedState;
var defaultConfig = {
  idProp: 'id',
  updateOnMount: false
};
/**
 * Adds shared state to a component.
 *
 * The purpose of shared state is to store framework component state at significant container
 * boundaries in order to restore it when the "same" component is mounted later.
 *
 * "Sameness" is determined by the `idProp` config member (defaults to "id"). If multiple
 * descendants have the same `idProp` within the subtree, SharedStateDecorator will not distinguish
 * between them and will allow each to read from and write over each other's data.
 *
 * For example, Panels and Panel are considered "significant container boundaries" since they are
 * key building blocks for sandstone applications. When components are rendered within a Panel, we
 * may want to store those components state on unmount so that we can restore it when returning to
 * the panel. Panel can (and does) use SharedStateDecorator to establish a shared state which can be
 * used by contained components.
 *
 * It's important to note that SharedStateDecorator doesn't prescribe how or what is stored nor how
 * the data is managed. That is left to the consuming component to determine. Also, unlike React
 * state or third-party state management solutions like Redux, updating shared state will not
 * initiate an update cycle in React. The intent is only to restore state on mount.
 *
 * If shared state is used in the render method for a component, it may be necessary to use the
 * `updateOnMount` config member which will initiate an update cycle within React once the data is
 * available from an upstream shared state.
 *
 * @hoc
 * @private
 */

var SharedStateDecorator = (0, _hoc["default"])(defaultConfig, function (config, Wrapped) {
  var _class;

  var idProp = config.idProp,
      updateOnMount = config.updateOnMount;
  return _class = /*#__PURE__*/function (_Component) {
    _inherits(_class, _Component);

    var _super = _createSuper(_class);

    function _class(props) {
      var _this;

      _classCallCheck(this, _class);

      _this = _super.call(this, props);
      _this.data = {};
      _this.sharedState = _this.initSharedState();
      _this.state = {
        updateOnMount: false
      };
      return _this;
    }

    _createClass(_class, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        this.loadFromContext();
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate(prevProps) {
        if (!prevProps.noSharedState && this.props.noSharedState) {
          this.data = {};
        } else if (prevProps.noSharedState && !this.props.noSharedState) {
          this.loadFromContext();
        }
      }
    }, {
      key: "isUpdateable",
      value: function isUpdateable() {
        var _this$props = this.props,
            id = _this$props[idProp],
            noSharedState = _this$props.noSharedState;
        return !noSharedState && (id || id === 0);
      }
    }, {
      key: "initSharedState",
      value: function initSharedState() {
        var _this2 = this;

        return {
          set: function set(key, value) {
            var id = _this2.props[idProp];

            if (_this2.isUpdateable()) {
              _this2.data[id] = _this2.data[id] || {};
              _this2.data[id][key] = value;
            }
          },
          get: function get(key) {
            var id = _this2.props[idProp];
            return _this2.isUpdateable() && _this2.data[id] ? _this2.data[id][key] : null;
          },
          "delete": function _delete(key) {
            var id = _this2.props[idProp];

            if (_this2.isUpdateable() && _this2.data[id]) {
              delete _this2.data[id][key];
            }
          }
        };
      }
    }, {
      key: "loadFromContext",
      value: function loadFromContext() {
        var _this$props2 = this.props,
            id = _this$props2[idProp],
            noSharedState = _this$props2.noSharedState;

        if (!noSharedState && this.context && this.context.get) {
          var data = this.context.get(id);

          if (data) {
            this.data = data;
          } else {
            this.context.set(id, this.data);
          }

          if (updateOnMount) {
            this.setState({
              updateOnMount: true
            });
          }
        }
      }
    }, {
      key: "render",
      value: function render() {
        var props = Object.assign({}, this.props);
        delete props.noSharedState;
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(SharedState.Provider, {
          value: this.sharedState,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread({}, props))
        });
      }
    }]);

    return _class;
  }(_react.Component), _class.displayName = 'SharedStateDecorator', _class.contextType = SharedState, _class.propTypes = {
    /**
     * Prevents the component from setting or restoring any framework shared state.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    noSharedState: _propTypes["default"].bool
  }, _class;
});
exports.SharedStateDecorator = SharedStateDecorator;
var _default = SharedStateDecorator;
exports["default"] = _default;
