// @enact/sandstone/Spinner/Spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SpinnerDecorator = exports.SpinnerBase = exports.Spinner = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));

var _compose = _interopRequireDefault(require("ramda/src/compose"));

var _react = require("react");

var _Pause = _interopRequireDefault(require("@enact/spotlight/Pause"));

var _Spinner = _interopRequireDefault(require("@enact/ui/Spinner"));

var _spotlight = _interopRequireDefault(require("@enact/spotlight"));

var _$L = _interopRequireDefault(require("../internal/$L"));

var _Marquee = _interopRequireDefault(require("../Marquee"));

var _Skinnable = _interopRequireDefault(require("../Skinnable"));

var _SpinnerModule = _interopRequireDefault(require("./Spinner.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["children", "css"],
    _excluded2 = ["children", "css"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A component that shows spinning balls, with optional text as children.
 *
 * @class SpinnerCore
 * @memberof sandstone/Spinner
 * @ui
 * @private
 */
var SpinnerCore = (0, _kind["default"])({
  name: 'SpinnerCore',
  propTypes: {
    css: _propTypes["default"].object
  },
  styles: {
    css: _SpinnerModule["default"]
  },
  computed: {
    'aria-label': function ariaLabel(_ref) {
      var aria = _ref['aria-label'],
          children = _ref.children;

      if (aria) {
        return aria;
      } else if (!children) {
        return (0, _$L["default"])('Loading');
      }
    }
  },
  render: function render(_ref2) {
    var children = _ref2.children,
        css = _ref2.css,
        rest = _objectWithoutProperties(_ref2, _excluded);

    return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", _objectSpread(_objectSpread({
      "aria-live": "off",
      role: "alert"
    }, rest), {}, {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: css.bg,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: css.decorator,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: css.fan1
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: css.fan2
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: css.fan3
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: css.fan4
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: css.cap
          })]
        })
      }), children ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_Marquee["default"], {
        className: css.client,
        marqueeOn: "render",
        alignment: "center",
        children: children
      }) : null]
    }));
  }
});
/**
 * The base component, defining all of the properties.
 *
 * @class SpinnerBase
 * @memberof sandstone/Spinner
 * @extends ui/Spinner.SpinnerBase
 * @ui
 * @public
 */

var SpinnerBase = (0, _kind["default"])({
  name: 'Spinner',
  propTypes:
  /** @lends sandstone/Spinner.SpinnerBase.prototype */
  {
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `spinner` - The root component class, unless there is a scrim. The scrim and floating
     *	layer can be a sibling or parent to this root "spinner" element.
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object,

    /**
     * Customize the size of this component.
     *
     * Recommended usage is "medium" (default) for standalone and popup scenarios, while "small"
     * is best suited for use inside other elements, like {@link sandstone/Item.Item}.
     *
     * @type {('medium'|'small')}
     * @default 'medium'
     * @public
     */
    size: _propTypes["default"].oneOf(['medium', 'small']),

    /**
     * Removes the background color (making it transparent).
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    transparent: _propTypes["default"].bool
  },
  defaultProps: {
    size: 'medium',
    transparent: false
  },
  styles: {
    css: _SpinnerModule["default"],
    publicClassNames: 'spinner'
  },
  computed: {
    className: function className(_ref3) {
      var children = _ref3.children,
          size = _ref3.size,
          transparent = _ref3.transparent,
          styler = _ref3.styler;
      return styler.append(size, {
        content: !!children,
        transparent: transparent
      });
    }
  },
  render: function render(_ref4) {
    var children = _ref4.children,
        css = _ref4.css,
        rest = _objectWithoutProperties(_ref4, _excluded2);

    delete rest.transparent;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Spinner["default"], _objectSpread(_objectSpread({}, rest), {}, {
      css: css,
      component: SpinnerCore,
      children: children
    }));
  }
});
/**
 * A higher-order component that pauses spotlight when `blockClickOn` prop is `'screen'`.
 *
 * Resumes spotlight when unmounted. However, spotlight is not paused when `blockClickOn` prop is
 * `'container'`. Blocking spotlight within the container is up to app implementation.
 *
 * @hoc
 * @memberof sandstone/Spinner
 * @private
 */

exports.SpinnerBase = SpinnerBase;
var SpinnerSpotlightDecorator = (0, _hoc["default"])(function (config, Wrapped) {
  var _class;

  return _class = /*#__PURE__*/function (_Component) {
    _inherits(_class, _Component);

    var _super = _createSuper(_class);

    function _class(props) {
      var _this;

      _classCallCheck(this, _class);

      _this = _super.call(this, props);
      _this.paused = new _Pause["default"]('Spinner');
      var blockClickOn = props.blockClickOn;

      var current = _spotlight["default"].getCurrent();

      if (blockClickOn === 'screen') {
        _this.paused.pause();

        if (current) {
          current.blur();
        }
      }

      return _this;
    }

    _createClass(_class, [{
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        var blockClickOn = this.props.blockClickOn;

        if (blockClickOn === 'screen') {
          _spotlight["default"].focus();

          this.paused.resume();
        }
      }
    }, {
      key: "render",
      value: function render() {
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread({}, this.props));
      }
    }]);

    return _class;
  }(_react.Component), _class.displayName = 'SpinnerSpotlightDecorator', _class.propTypes =
  /** @lends sandstone/Spinner.Spinner.prototype */
  {
    /**
     * Determines how far the click-blocking should extend.
     *
     * It can be either `'screen'`, `'container'`, or `null`. `'screen'` pauses spotlight.
     * Changing this property to `'screen'` after creation is not supported.
     *
     * @type {('screen'|'container')}
     * @default null
     * @public
     */
    blockClickOn: _propTypes["default"].oneOf(['screen', 'container', null])
  }, _class;
});
/**
 * Sandstone-specific Spinner behaviors to apply to [Spinner]{@link sandstone/Spinner.Spinner}.
 *
 * @hoc
 * @memberof sandstone/Spinner
 * @mixes sandstone/Skinnable.Skinnable
 * @public
 */

var SpinnerDecorator = (0, _compose["default"])(_Pure["default"], SpinnerSpotlightDecorator, _Skinnable["default"]);
/**
 * A Sandstone-styled Spinner.
 *
 * @class Spinner
 * @memberof sandstone/Spinner
 * @extends sandstone/Spinner.SpinnerBase
 * @mixes sandstone/Spinner.SpinnerDecorator
 * @ui
 * @public
 */

exports.SpinnerDecorator = SpinnerDecorator;
var Spinner = SpinnerDecorator(SpinnerBase);
exports.Spinner = Spinner;
var _default = Spinner;
exports["default"] = _default;
