// @enact/core/internal/Registry/Registry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.Registry = void 0;

/**
 * Allows components to register parents to cascade context changes and trigger functions
 *
 * @type Object
 * @memberof core/Registry
 * @private
 */
var Registry = {
  create: function create(handler) {
    var instances = [];
    var currentParent;
    var registry = Object.freeze({
      set parent(register) {
        if (currentParent && currentParent.unregister) {
          currentParent.unregister();
        }

        if (register && typeof register === 'function') {
          currentParent = register(registry.notify);
        }
      },

      notify: function notify(ev) {
        var exclude = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : function () {
          return true;
        };
        instances.filter(exclude).forEach(function (f) {
          return f(ev);
        });
      },
      register: function register(instance) {
        if (instances.indexOf(instance) === -1) {
          instances.push(instance);

          if (handler) {
            handler({
              action: 'register'
            }, instance);
          }
        }

        return {
          notify: function notify(ev) {
            if (handler) {
              handler(ev, instance);
            }
          },
          unregister: function unregister() {
            var i = instances.indexOf(instance);

            if (i >= 0) {
              instances.splice(i, 1);

              if (handler) {
                handler({
                  action: 'register'
                }, instance);
              }
            }
          }
        };
      }
    });
    return registry;
  }
};
exports.Registry = Registry;
var _default = Registry;
exports["default"] = _default;
