// @enact/spotlight/SpotlightContainerDecorator/useSpotlightContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
exports.useSpotlightContainer = useSpotlightContainer;

var _useClass = _interopRequireDefault(require("@enact/core/useClass"));

var _react = require("react");

var _SpotlightContainer = _interopRequireDefault(require("./SpotlightContainer"));

var _excluded = ["disabled", "id", "muted", "restrict"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Configuration for `useSpotlightContainer`
 *
 * @typedef {Object} useSpotlightContainerConfig
 * @memberof spotlight/SpotlightContainerDecorator
 * @property {Object}   [containerConfig]                   Spotlight container configuration. This
 *                                                          is a design-time configuration and only
 *                                                          applied once on first render.
 * @property {Function} [navigableFilter]                   Called to determine if a spottable
 *                                                          component is accessible. This is a
 *                                                          design-time configuration and only
 *                                                          applied once on first render.
 * @property {Boolean}  [preserveId]                        Preserves the spotlight `id`, if
 *                                                          specified, when the component unmounts.
 * @property {Boolean}  [disabled]                          Prevents controls in the container from
 *                                                          being navigable. This is a design-time
 *                                                          configuration and only applied once on
 *                                                          first render.
 * @property {Boolean}  [muted]                             Suppresses appearance of focus within
 *                                                          the container while still allowing the
 *                                                          component to gain focus. The components
 *                                                          within the container are responsible for
 *                                                          adjusting their visual appearance based
 *                                                          on the presence of the
 *                                                          `data-spotlight-container-muted`
 *                                                          attribute on an ancestor container.
 * @property {String}   [id]                                Identifies this component within the
 *                                                          Spotlight system. If unset, an id will
 *                                                          be generated.
 * @property {('self-only'|'self-first'|'none')} [restrict] Restricts or prioritizes navigation when
 *                                                          focus attempts to leave the container.
 *                                                          It can be either 'none', 'self-first',
 *                                                          or 'self-only'. Specifying 'self-first'
 *                                                          indicates that elements within the
 *                                                          container will have a higher likelihood
 *                                                          to be chosen as the next navigable
 *                                                          element. Specifying 'self-only'
 *                                                          indicates that elements in other
 *                                                          containers cannot be navigated to by
 *                                                          using 5-way navigation - however,
 *                                                          elements in other containers can still
 *                                                          receive focus by calling
 *                                                          `Spotlight.focus(elem)` explicitly.
 *                                                          Specifying 'none' indicates there should
 *                                                          be no restrictions when 5-way navigating
 *                                                          the container.
 * @private
 */

/**
 * Object returned by `useSpotlightContainer`.
 *
 * @typedef {Object} useSpotlightContainerInterface
 * @memberof spotlight/SpotlightContainerDecorator
 * @property {Function} onBlurCapture    Callback to notify hook when an node was blurred within the
 *                                       container. Must be called during the capture phase of the
 *                                       event flow.
 * @property {Function} onFocusCapture   Callback to notify hook when an node was focused within the
 *                                       container. Must be called during the capture phase of the
 *                                       event flow.
 * @property {String}   attributes       An object of DOM attributes of representing the current
 *                                       spotlight container metadata. Must be passed onto the
 *                                       container node.
 * @property {Function} onPointerEnter   Callback to notify hook when the pointer enters the
 *                                       container. Must be called during the capture phase of the
 *                                       event flow.
 * @property {Function} onPointerLeave   Callback to notify hook when the pointer leaves the
 *                                       container. Must be called during the capture phase of the
 *                                       event flow.
 * @private
 */

/**
 * Support Spotlight container
 *
 * @param {useSpotlightContainerConfig} [config] Configuration options
 * @returns {useSpotlightContainerInterface}
 * @private
 */
function useSpotlightContainer() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var disabled = config.disabled,
      id = config.id,
      muted = config.muted,
      restrict = config.restrict,
      rest = _objectWithoutProperties(config, _excluded);

  var spotlightContainer = (0, _useClass["default"])(_SpotlightContainer["default"], rest);
  spotlightContainer.setProps({
    disabled: disabled,
    id: id,
    muted: muted,
    restrict: restrict
  });
  (0, _react.useLayoutEffect)(function () {
    return function () {
      spotlightContainer.unload();
    };
  }, [spotlightContainer]);
  return {
    attributes: spotlightContainer.attributes,
    onBlurCapture: spotlightContainer.onBlurCapture,
    onFocusCapture: spotlightContainer.onFocusCapture,
    onPointerEnter: spotlightContainer.onPointerEnter,
    onPointerLeave: spotlightContainer.onPointerLeave
  };
}

var _default = useSpotlightContainer;
exports["default"] = _default;
