// @enact/ui/useScroll/utilDOM.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.utilDOM = exports["default"] = void 0;

var _warning = _interopRequireDefault(require("warning"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// At the end, wes should not use DOM APIs as well as the APIs in the `utilDOM`. If we use them, we have to try to remove them first if possible.
var utilDOM = function () {
  // Functions
  function containsDangerously(ref, target) {
    if (!target) {
      return false;
    } else if (ref.current) {
      process.env.NODE_ENV !== "production" ? (0, _warning["default"])(ref.current.contains, 'The `contains` function of the Ref is not supported.') : void 0;
      return ref.current.contains(target);
    } else if (ref) {
      process.env.NODE_ENV !== "production" ? (0, _warning["default"])(ref.contains, 'The `contains` function of the Ref is not supported.') : void 0;
      return ref.contains(target);
    }

    return false;
  } // TBD
  // Need to define functions for the following DOM APIs
  // getBoundingClientRect
  // dataset.spotlightId
  // getRect
  // setAttribute
  // removeAttribute
  // scrollLeft
  // scrollTop


  function datasetDangerously() {}

  function getBoundingClientRectDangerously() {}

  function getRectDangerously() {}

  function removeAttributeDangerously() {}

  function setAttributeDangerously() {}

  var scrollLeftDangerously = 0;
  var scrollTopDangerously = 0; // Return

  return {
    containsDangerously: containsDangerously,
    datasetDangerously: datasetDangerously,
    getBoundingClientRectDangerously: getBoundingClientRectDangerously,
    getRectDangerously: getRectDangerously,
    removeAttributeDangerously: removeAttributeDangerously,
    scrollLeftDangerously: scrollLeftDangerously,
    scrollTopDangerously: scrollTopDangerously,
    setAttributeDangerously: setAttributeDangerously
  };
}();

exports.utilDOM = utilDOM;
var _default = utilDOM;
exports["default"] = _default;
