// @enact/sandstone/internal/Picker/PickerButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.PickerButtonBase = exports.PickerButton = void 0;

var _handle = require("@enact/core/handle");

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _MarqueeController = require("@enact/ui/Marquee/MarqueeController");

var _Pure = _interopRequireDefault(require("@enact/ui/internal/Pure"));

var _Touchable = _interopRequireDefault(require("@enact/ui/Touchable"));

var _Icon = _interopRequireDefault(require("../../Icon"));

var _Button = _interopRequireDefault(require("../../Button"));

var _PickerModule = _interopRequireDefault(require("./Picker.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["disabled", "icon"],
    _excluded2 = ["disabled", "icon", "joined"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var JoinedPickerButtonBase = (0, _kind["default"])({
  name: 'JoinedPickerButtonBase',
  propTypes: {
    disabled: _propTypes["default"].bool,
    icon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object])
  },
  render: function render(_ref) {
    var disabled = _ref.disabled,
        icon = _ref.icon,
        rest = _objectWithoutProperties(_ref, _excluded);

    return /*#__PURE__*/(0, _jsxRuntime.jsx)("span", _objectSpread(_objectSpread({}, rest), {}, {
      "data-webos-voice-intent": "Select",
      disabled: disabled,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icon["default"], {
        className: _PickerModule["default"].icon,
        disabled: disabled,
        size: "small",
        children: icon
      })
    }));
  }
});
var JoinedPickerButton = (0, _Touchable["default"])(JoinedPickerButtonBase);
var PickerButtonBase = (0, _kind["default"])({
  name: 'PickerButton',
  contextType: _MarqueeController.MarqueeControllerContext,
  propTypes: {
    disabled: _propTypes["default"].bool,
    icon: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object]),
    joined: _propTypes["default"].bool,
    onSpotlightDisappear: _propTypes["default"].func,
    spotlightDisabled: _propTypes["default"].bool
  },
  styles: {
    css: _PickerModule["default"]
  },
  handlers: {
    onMouseEnter: (0, _handle.handle)((0, _handle.forward)('onMouseEnter'), function (ev, props, sync) {
      if (sync && sync.enter) {
        sync.enter(null);
      }
    }),
    onMouseLeave: (0, _handle.handle)((0, _handle.forward)('onMouseLeave'), function (ev, props, sync) {
      if (sync && sync.leave) {
        sync.leave(null);
      }
    })
  },
  render: function render(_ref2) {
    var disabled = _ref2.disabled,
        icon = _ref2.icon,
        joined = _ref2.joined,
        rest = _objectWithoutProperties(_ref2, _excluded2);

    if (joined) {
      delete rest.onSpotlightDisappear;
      delete rest.spotlightDisabled;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(JoinedPickerButton, _objectSpread(_objectSpread({}, rest), {}, {
        icon: icon,
        disabled: disabled
      }));
    } else {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Button["default"], _objectSpread(_objectSpread({}, rest), {}, {
        backgroundOpacity: "transparent",
        disabled: disabled,
        icon: icon,
        size: "small"
      }));
    }
  }
});
exports.PickerButtonBase = PickerButtonBase;
var PickerButton = (0, _Pure["default"])(PickerButtonBase);
exports.PickerButton = PickerButton;
var _default = PickerButton;
exports["default"] = _default;
