// @enact/sandstone/internal/DateComponentPicker/DateComponentRangePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.DateComponentRangePickerBase = exports.DateComponentRangePicker = void 0;

var _Changeable = _interopRequireDefault(require("@enact/ui/Changeable"));

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _RangePicker = _interopRequireDefault(require("../../RangePicker"));

var _DateComponentPickerModule = _interopRequireDefault(require("./DateComponentPicker.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["accessibilityHint", "label", "max", "min", "noAnimation", "value", "wrap", "voiceLabel"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * {@link sandstone/internal/DataComponentPicker.DateComponentRangePicker} allows the selection of
 * one part of the date or time using a {@link sandstone/RangePicker.RangePicker}.
 *
 * @class DateComponentRangePickerBase
 * @memberof sandstone/internal/DateComponentPicker
 * @ui
 * @private
 */
var DateComponentRangePickerBase = (0, _kind["default"])({
  name: 'DateComponentRangePicker',
  propTypes:
  /** @lends sandstone/internal/DateComponentPicker.DateComponentRangePickerBase.prototype */
  {
    /**
     * The maximum value for the date component
     *
     * @type {Number}
     * @required
     */
    max: _propTypes["default"].number.isRequired,

    /**
     * The minimum value for the date component
     *
     * @type {Number}
     * @required
     */
    min: _propTypes["default"].number.isRequired,

    /**
     * The value of the date component
     *
     * @type {Number}
     * @required
     */
    value: _propTypes["default"].number.isRequired,

    /**
     * Sets the hint string read when focusing the picker.
     *
     * @type {String}
     * @public
     */
    accessibilityHint: _propTypes["default"].string,

    /**
     * The label to display below the picker
     *
     * @type {String}
     */
    label: _propTypes["default"].string,

    /**
     * By default, the picker will animate transitions between items if it has a defined
     * `width`. Specifying `noAnimation` will prevent any transition animation for the
     * component.
     *
     * @type {Boolean}
     * @public
     */
    noAnimation: _propTypes["default"].bool,

    /*
     * When `true`, allow the picker to continue from the opposite end of the list of options.
     *
     * @type {Boolean}
     * @public
     */
    wrap: _propTypes["default"].bool
  },
  styles: {
    css: _DateComponentPickerModule["default"],
    className: 'dateComponentPicker'
  },
  computed: {
    voiceLabel: function voiceLabel(_ref) {
      var min = _ref.min,
          max = _ref.max;
      return JSON.stringify([min, max]);
    }
  },
  render: function render(_ref2) {
    var accessibilityHint = _ref2.accessibilityHint,
        label = _ref2.label,
        max = _ref2.max,
        min = _ref2.min,
        noAnimation = _ref2.noAnimation,
        value = _ref2.value,
        wrap = _ref2.wrap,
        voiceLabel = _ref2.voiceLabel,
        rest = _objectWithoutProperties(_ref2, _excluded);

    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_RangePicker["default"], _objectSpread(_objectSpread({}, rest), {}, {
      accessibilityHint: accessibilityHint == null ? label : accessibilityHint,
      "data-webos-voice-labels-ext": voiceLabel,
      joined: true,
      max: max,
      min: min,
      noAnimation: noAnimation,
      orientation: "vertical",
      value: value,
      wrap: wrap
    }));
  }
});
/**
 * {@link sandstone/internal/DateComponentPicker.DateComponentRangePicker} allows the selection of one
 * part of the date (date, month, or year). It is a stateful component but allows updates by
 * providing a new `value` via props.
 *
 * @class DateComponentRangePicker
 * @memberof sandstone/internal/DateComponentPicker
 * @mixes ui/Changeable.Changeable
 * @ui
 * @private
 */

exports.DateComponentRangePickerBase = DateComponentRangePickerBase;
var DateComponentRangePicker = (0, _Changeable["default"])(DateComponentRangePickerBase);
exports.DateComponentRangePicker = DateComponentRangePicker;
var _default = DateComponentRangePicker;
exports["default"] = _default;
