// @enact/ui/FloatingLayer/FloatingLayerDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FloatingLayerContext", {
  enumerable: true,
  get: function get() {
    return _useFloatingLayerDecorator.FloatingLayerContext;
  }
});
exports["default"] = exports.FloatingLayerDecorator = void 0;
Object.defineProperty(exports, "useFloatingLayerDecorator", {
  enumerable: true,
  get: function get() {
    return _useFloatingLayerDecorator.useFloatingLayerDecorator;
  }
});

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _useFloatingLayerDecorator = require("./useFloatingLayerDecorator");

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["className"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Default config for {@link ui/FloatingLayer.FloatingLayerDecorator}.
 *
 * @memberof ui/FloatingLayer.FloatingLayerDecorator
 * @hocconfig
 */
var defaultConfig = {
  /**
   * Element Id of the floatLayer
   *
   * @type {String}
   * @default 'floatLayer'
   * @public
   * @memberof ui/FloatingLayer.FloatingLayerDecorator.defaultConfig
   */
  floatLayerId: 'floatLayer',

  /**
   * Class name to be applied to wrapped component.
  	 * It can be used when you want to only apply
   * certain styles to the wrapped component and not to the float layer.
   *
   * @type {String}
   * @default ''
   * @public
   * @memberof ui/FloatingLayer.FloatingLayerDecorator.defaultConfig
   */
  wrappedClassName: ''
};
/**
 * A higher-order component that adds a FloatingLayer adjacent to the wrapped component.
 *
 * Any classes passed to the FloatingLayerDecorator will be applied to a wrapper surrounding the
 * wrapped component and the floating layer. If specified in the config, `wrappedClassName` will be
 * passed as `className` to the wrapped component.
 *
 * @class FloatingLayerDecorator
 * @memberof ui/FloatingLayer
 * @hoc
 * @public
 */

var FloatingLayerDecorator = (0, _hoc["default"])(defaultConfig, function (config, Wrapped) {
  var floatLayerId = config.floatLayerId,
      wrappedClassName = config.wrappedClassName; // eslint-disable-next-line no-shadow

  function FloatingLayerDecorator(props) {
    var className = props.className,
        rest = _objectWithoutProperties(props, _excluded);

    var hook = (0, _useFloatingLayerDecorator.useFloatingLayerDecorator)({
      className: className,
      floatLayerId: floatLayerId
    });
    return hook.provideFloatingLayer( /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread(_objectSpread({}, rest), {}, {
      className: wrappedClassName
    }), "floatWrapped"));
  }

  return FloatingLayerDecorator;
});
exports.FloatingLayerDecorator = FloatingLayerDecorator;
var _default = FloatingLayerDecorator;
exports["default"] = _default;
