// @enact/sandstone/Switch/Switch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SwitchDecorator = exports.SwitchBase = exports.Switch = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _compose = _interopRequireDefault(require("ramda/src/compose"));

var _Spottable = _interopRequireDefault(require("@enact/spotlight/Spottable"));

var _Touchable = _interopRequireDefault(require("@enact/ui/Touchable"));

var _Toggleable = _interopRequireDefault(require("@enact/ui/Toggleable"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _Skinnable = _interopRequireDefault(require("../Skinnable"));

var _SwitchModule = _interopRequireDefault(require("./Switch.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["children", "css", "disabled", "selected"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Renders the base level DOM structure of the component.
 *
 * @class SwitchBase
 * @memberof sandstone/Switch
 * @extends sandstone/Icon.Icon
 * @ui
 * @public
 */
var SwitchBase = (0, _kind["default"])({
  name: 'Switch',
  propTypes:
  /** @lends sandstone/Switch.SwitchBase.prototype */
  {
    children: _propTypes["default"].string,
    css: _propTypes["default"].object,

    /**
     * Disables Switch and becomes non-interactive.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    disabled: _propTypes["default"].bool,

    /**
     * Disables animation.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    noAnimation: _propTypes["default"].bool,

    /**
     * Sets whether this control is in the 'on' or 'off' state. `true` for 'on', `false` for 'off'.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selected: _propTypes["default"].bool
  },
  defaultProps: {
    children: 'circle',
    noAnimation: false,
    selected: false
  },
  styles: {
    css: _SwitchModule["default"],
    className: 'switch',
    publicClassNames: ['switch', 'selected', 'client']
  },
  computed: {
    className: function className(_ref) {
      var noAnimation = _ref.noAnimation,
          selected = _ref.selected,
          styler = _ref.styler;
      return styler.append({
        animated: !noAnimation,
        selected: selected
      });
    }
  },
  render: function render(_ref2) {
    var children = _ref2.children,
        css = _ref2.css,
        disabled = _ref2.disabled,
        selected = _ref2.selected,
        rest = _objectWithoutProperties(_ref2, _excluded);

    delete rest.noAnimation;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", _objectSpread(_objectSpread({}, rest), {}, {
      "aria-pressed": selected,
      "aria-disabled": disabled,
      disabled: disabled,
      role: "button",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: css.bg
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: css.client,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icon["default"], {
          size: "small",
          className: css.icon,
          children: children
        })
      })]
    }));
  }
});
/**
 * Adds interactive functionality to `Switch`.
 *
 * @class SwitchDecorator
 * @memberof sandstone/Switch
 * @mixes ui/Toggleable.Toggleable
 * @mixes spotlight/Spottable.Spottable
 * @hoc
 * @public
 */

exports.SwitchBase = SwitchBase;
var SwitchDecorator = (0, _compose["default"])((0, _Toggleable["default"])({
  toggleProp: 'onClick'
}), _Touchable["default"], _Spottable["default"], _Skinnable["default"]);
/**
 * A Sandstone-styled component that looks like a toggle switch.
 *
 * `Switch` will manage its `selected` state via [Toggleable]{@link ui/Toggleable} unless set
 * directly.
 *
 * @class Switch
 * @memberof sandstone/Switch
 * @extends sandstone/Switch.SwitchBase
 * @mixes sandstone/Switch.SwitchDecorator
 * @ui
 * @public
 */

exports.SwitchDecorator = SwitchDecorator;
var Switch = SwitchDecorator(SwitchBase);
exports.Switch = Switch;
var _default = Switch;
exports["default"] = _default;
