// @enact/sandstone/internal/Panels/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AutoFocusDecorator", {
  enumerable: true,
  get: function get() {
    return _useAutoFocus.AutoFocusDecorator;
  }
});
Object.defineProperty(exports, "BasicArranger", {
  enumerable: true,
  get: function get() {
    return _Arrangers.BasicArranger;
  }
});
Object.defineProperty(exports, "CancelDecorator", {
  enumerable: true,
  get: function get() {
    return _CancelDecorator["default"];
  }
});
Object.defineProperty(exports, "CrossFadeArranger", {
  enumerable: true,
  get: function get() {
    return _Arrangers.CrossFadeArranger;
  }
});
Object.defineProperty(exports, "FadeAndSlideArranger", {
  enumerable: true,
  get: function get() {
    return _Arrangers.FadeAndSlideArranger;
  }
});
Object.defineProperty(exports, "FloatingLayerIdProvider", {
  enumerable: true,
  get: function get() {
    return _FloatingLayerIdProvider["default"];
  }
});
Object.defineProperty(exports, "NavigationButton", {
  enumerable: true,
  get: function get() {
    return _NavigationButton["default"];
  }
});
Object.defineProperty(exports, "PanelsStateContext", {
  enumerable: true,
  get: function get() {
    return _Viewport.PanelsStateContext;
  }
});
Object.defineProperty(exports, "PopupDecorator", {
  enumerable: true,
  get: function get() {
    return _PopupDecorator["default"];
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function get() {
    return _Viewport["default"];
  }
});
Object.defineProperty(exports, "useAutoFocus", {
  enumerable: true,
  get: function get() {
    return _useAutoFocus["default"];
  }
});

var _Arrangers = require("./Arrangers");

var _CancelDecorator = _interopRequireDefault(require("./CancelDecorator"));

var _FloatingLayerIdProvider = _interopRequireDefault(require("./FloatingLayerIdProvider"));

var _NavigationButton = _interopRequireDefault(require("./NavigationButton"));

var _PopupDecorator = _interopRequireDefault(require("./PopupDecorator"));

var _Viewport = _interopRequireWildcard(require("./Viewport"));

var _useAutoFocus = _interopRequireWildcard(require("./useAutoFocus"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
