// @enact/sandstone/MediaPlayer/Times.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.TimesBase = exports.Times = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _util = require("../internal/util");

var _util2 = require("./util");

var _TimesModule = _interopRequireDefault(require("./Times.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["currentPeriod", "currentReadable", "noCurrentTime", "noSeparator", "noTotalTime", "totalPeriod", "totalReadable"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Sandstone-styled formatted time component.
 *
 * @class Times
 * @memberof sandstone/MediaPlayer
 * @ui
 * @public
 */
var TimesBase = (0, _kind["default"])({
  name: 'Times',
  propTypes:
  /** @lends sandstone/MediaPlayer.Times.prototype */
  {
    /**
     * An instance of a Duration Formatter from i18n.
     *
     * Must has a `format()` method that returns a string.
     *
     * @type {Object}
     * @required
     * @public
     */
    formatter: _propTypes["default"].object.isRequired,

    /**
     * The current time in seconds of the video source.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    current: _propTypes["default"].number,

    /**
     * Removes the current time.
     *
     * @type {Boolean}
     * @public
     */
    noCurrentTime: _propTypes["default"].bool,

    /**
     * Removes the total time.
     *
     * @type {Boolean}
     * @public
     */
    noTotalTime: _propTypes["default"].bool,

    /**
     * The total time (duration) in seconds of the loaded video source.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    total: _propTypes["default"].number
  },
  defaultProps: {
    current: 0,
    total: 0
  },
  styles: {
    css: _TimesModule["default"],
    className: 'times'
  },
  computed: {
    currentPeriod: function currentPeriod(_ref) {
      var current = _ref.current;
      return (0, _util2.secondsToPeriod)(current);
    },
    currentReadable: function currentReadable(_ref2) {
      var current = _ref2.current,
          formatter = _ref2.formatter;
      return (0, _util2.secondsToTime)(current, formatter);
    },
    noSeparator: function noSeparator(_ref3) {
      var noCurrentTime = _ref3.noCurrentTime,
          noTotalTime = _ref3.noTotalTime;
      return noCurrentTime || noTotalTime;
    },
    totalPeriod: function totalPeriod(_ref4) {
      var total = _ref4.total;
      return (0, _util2.secondsToPeriod)(total);
    },
    totalReadable: function totalReadable(_ref5) {
      var total = _ref5.total,
          formatter = _ref5.formatter;
      return (0, _util2.secondsToTime)(total, formatter);
    }
  },
  render: function render(_ref6) {
    var currentPeriod = _ref6.currentPeriod,
        currentReadable = _ref6.currentReadable,
        noCurrentTime = _ref6.noCurrentTime,
        noSeparator = _ref6.noSeparator,
        noTotalTime = _ref6.noTotalTime,
        totalPeriod = _ref6.totalPeriod,
        totalReadable = _ref6.totalReadable,
        rest = _objectWithoutProperties(_ref6, _excluded);

    delete rest.current;
    delete rest.formatter;
    delete rest.total;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", _objectSpread(_objectSpread({}, rest), {}, {
      children: [noCurrentTime ? null : /*#__PURE__*/(0, _jsxRuntime.jsx)("time", {
        className: _TimesModule["default"].currentTime,
        dateTime: currentPeriod,
        children: currentReadable
      }), noSeparator ? null : /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        className: _TimesModule["default"].separator,
        children: "/"
      }), noTotalTime ? null : /*#__PURE__*/(0, _jsxRuntime.jsx)("time", {
        className: _TimesModule["default"].totalTime,
        dateTime: totalPeriod,
        children: totalReadable
      })]
    }));
  }
});
exports.TimesBase = TimesBase;
var Times = (0, _util.onlyUpdateForProps)(TimesBase, ['current', 'formatter', 'total']);
exports.Times = Times;
var _default = Times;
exports["default"] = _default;
