// @enact/spotlight/src/tests/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testScenario = exports.spottable = exports.someSpottablesAndContainers = exports.someSpottables = exports.someNodes = exports.someContainers = exports.node = exports.join = exports.generateContainerId = exports.container = void 0;

var _ramda = _interopRequireDefault(require("ramda"));

var _container = require("../container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var join = _ramda["default"].unapply(_ramda["default"].join('\n'));

exports.join = join;

var testScenario = function testScenario(scenario, callback) {
  return function () {
    var rootId = 'test-root';
    var html = join("<div id=\"".concat(rootId, "\">"), scenario, '</div>');
    document.body.innerHTML = html;
    var root = document.getElementById(rootId);

    try {
      callback(root);
    } catch (e) {
      console.log(html); // eslint-disable-line no-console

      throw e;
    }
  };
};

exports.testScenario = testScenario;
var _id = 1;

var generateContainerId = function generateContainerId() {
  return "".concat(_container.containerAttribute, "=").concat(_id++);
};

exports.generateContainerId = generateContainerId;

var coerceProps = function coerceProps(v) {
  if (typeof v === 'object') {
    return v;
  } else if (typeof v !== 'undefined') {
    return {
      children: v
    };
  }
};

var node = function node(props) {
  var children = '';
  var attributes = '';
  Object.keys(props).forEach(function (key) {
    if (key === 'children') {
      children = props.children;
    } else {
      var value = props[key];
      if (key === 'className') key = 'class';
      attributes += "".concat(key, "=\"").concat(value, "\" ");
    }
  });
  return "<div ".concat(attributes, ">").concat(children, "</div>");
};

exports.node = node;

var spottable = function spottable(props) {
  return node(_objectSpread({
    className: 'spottable'
  }, coerceProps(props)));
};

exports.spottable = spottable;

var container = function container(props) {
  var _objectSpread2;

  return node(_objectSpread((_objectSpread2 = {}, _defineProperty(_objectSpread2, _container.containerAttribute, _id++), _defineProperty(_objectSpread2, 'data-spotlight-container', true), _objectSpread2), coerceProps(props)));
}; // eslint-disable-next-line react-hooks/rules-of-hooks


exports.container = container;

var someNodes = _ramda["default"].useWith(_ramda["default"].compose(_ramda["default"].join('\n'), _ramda["default"].map), [_ramda["default"].identity, _ramda["default"].range(0)]);

exports.someNodes = someNodes;
var someSpottables = someNodes(spottable);
exports.someSpottables = someSpottables;
var someContainers = someNodes(container);
exports.someContainers = someContainers;

var someSpottablesAndContainers = _ramda["default"].converge(_ramda["default"].concat, [someSpottables, someContainers]);

exports.someSpottablesAndContainers = someSpottablesAndContainers;
