// @enact/ui/ImageItem/ImageItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.ImageItemDecorator = exports.ImageItemBase = exports.ImageItem = void 0;

var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes2 = _interopRequireDefault(require("prop-types"));

var _compose = _interopRequireDefault(require("ramda/src/compose"));

var _ComponentOverride = _interopRequireDefault(require("../ComponentOverride"));

var _ForwardRef = _interopRequireDefault(require("../ForwardRef"));

var _Image = _interopRequireDefault(require("../Image"));

var _Layout = require("../Layout");

var _ImageItemModule = _interopRequireDefault(require("./ImageItem.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["imageComponent"],
    _excluded2 = ["children", "componentRef", "css", "imageComponent", "orientation", "placeholder", "src"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

// Adapts ComponentOverride to work within Cell since both use the component prop
function ImageOverride(_ref) {
  var imageComponent = _ref.imageComponent,
      rest = _objectWithoutProperties(_ref, _excluded);

  return (0, _ComponentOverride["default"])(_objectSpread({
    component: imageComponent
  }, rest));
}
/**
 * A basic image item without any behavior.
 *
 * @class ImageItemBase
 * @memberof ui/ImageItem
 * @ui
 * @public
 */


var ImageItemBase = (0, _kind["default"])({
  name: 'ui:ImageItem',
  propTypes:
  /** @lends ui/ImageItem.ImageItemBase.prototype */
  {
    /**
     * The caption displayed with the image.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes2["default"].node,

    /**
     * Called with a reference to the root component.
     *
     * When using {@link ui/ImageItem.ImageItem}, the `ref` prop is forwarded to this component
     * as `componentRef`.
     *
     * @type {Object|Function}
     * @public
     */
    componentRef: _propTypes["default"].ref,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `imageItem` - The root component class
     * * `caption` - The caption component class
     * * `horizontal` - Applied when `orientation="horizontal"
     * * `image` - The image component class
     * * `selected` - Applied when `selected` prop is `true`
     * * `vertical` - Applied when `orientation="vertical"
     *
     * @type {Object}
     * @public
     */
    css: _propTypes2["default"].object,

    /**
     * The component used to render the image component.
     *
     * @type {Component|Element}
     * @public
     */
    imageComponent: _propTypes["default"].componentOverride,

    /**
     * The layout orientation of the component.
     *
     * @type {('horizontal'|'vertical')}
     * @default 'vertical'
     * @public
     */
    orientation: _propTypes2["default"].oneOf(['horizontal', 'vertical']),

    /**
     * A placeholder image to be displayed before the image is loaded.
     *
     * @type {String}
     * @public
     */
    placeholder: _propTypes2["default"].string,

    /**
     * Applies a selected visual effect to the image.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selected: _propTypes2["default"].bool,

    /**
     * String value or Object of values used to determine which image will appear on a specific
     * screenSize.
     *
     * @type {String|Object}
     * @public
     */
    src: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].object])
  },
  defaultProps: {
    imageComponent: _Image["default"],
    orientation: 'vertical',
    selected: false
  },
  styles: {
    css: _ImageItemModule["default"],
    className: 'imageItem',
    publicClassNames: true
  },
  computed: {
    className: function className(_ref2) {
      var orientation = _ref2.orientation,
          selected = _ref2.selected,
          styler = _ref2.styler;
      return styler.append({
        selected: selected,
        horizontal: orientation === 'horizontal',
        vertical: orientation === 'vertical'
      });
    }
  },
  render: function render(_ref3) {
    var children = _ref3.children,
        componentRef = _ref3.componentRef,
        css = _ref3.css,
        imageComponent = _ref3.imageComponent,
        orientation = _ref3.orientation,
        placeholder = _ref3.placeholder,
        src = _ref3.src,
        rest = _objectWithoutProperties(_ref3, _excluded2);

    delete rest.selected;
    var isHorizontal = orientation === 'horizontal';
    var Component = isHorizontal ? _Layout.Row : _Layout.Column;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Component, _objectSpread(_objectSpread({}, rest), {}, {
      ref: componentRef,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Layout.Cell, {
        className: css.image,
        component: ImageOverride,
        imageComponent: imageComponent,
        placeholder: placeholder,
        shrink: isHorizontal,
        src: src
      }), children ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_Layout.Cell, {
        className: css.caption,
        shrink: !isHorizontal // eslint-disable-next-line no-undefined
        ,
        align: isHorizontal ? 'center' : undefined,
        children: children
      }) : null]
    }));
  }
});
/**
 * A higher-order component that adds behaviors to an
 * [ImageItemBase]{@link ui/ImageItem.ImageItemBase}.
 *
 * @hoc
 * @memberof ui/ImageItem
 * @mixes ui/ForwardRef.ForwardRef
 * @public
 */

exports.ImageItemBase = ImageItemBase;
var ImageItemDecorator = (0, _compose["default"])((0, _ForwardRef["default"])({
  prop: 'componentRef'
}));
/**
 * A minimally styled ImageItem ready for customization by a theme.
 *
 * @class ImageItem
 * @memberof ui/ImageItem
 * @extends ui/ImageItem.ImageItemBase
 * @mixes ui/ImageItem.ImageItemDecorator
 * @omit componentRef
 * @ui
 * @public
 */

exports.ImageItemDecorator = ImageItemDecorator;
var ImageItem = ImageItemDecorator(ImageItemBase);
exports.ImageItem = ImageItem;
var _default = ImageItem;
exports["default"] = _default;
