// @enact/ui/Slider/Knob.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.Knob = void 0;

var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes2 = _interopRequireDefault(require("prop-types"));

var _ComponentOverride = _interopRequireDefault(require("../ComponentOverride"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["orientation", "proportion", "tooltipComponent", "value"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * An unstyled, knob component to be used with a `Slider`.
 *
 * @class Knob
 * @memberof ui/Slider
 * @ui
 * @public
 */
var Knob = (0, _kind["default"])({
  name: 'Knob',
  propTypes:
  /** @lends ui/Slider.Knob.prototype */
  {
    /**
     * The orientation of the slider, either `"horizontal"` or `"vertical"`.
     *
     * @type {String}
     * @default "horizontal"
     * @public
     */
    orientation: _propTypes2["default"].oneOf(['horizontal', 'vertical']),

    /**
     * A number between 0 and 1 representing the proportion of the `value` in terms of `min`
     * and `max` props of the slider.
     *
     * @type {Boolean}
     * @public
     */
    proportion: _propTypes2["default"].number,

    /**
     * Adds a tooltip to the knob using the provided component.
     *
     * The following props are forwarded to the tooltip:
     *
     * * `children` - The `value` prop from the slider
     * * `orientation` - The value of the `orientation` prop from the slider
     * * `proportion` - The value of the `proportion` prop from the knob
     *
     * This prop accepts either a Component (e.g. `MyTooltip`} which will be instantiated with
     * the above props or a component instance (e.g. `<MyTooltip customProp="value" />`) which
     * will have its props merged with the above props.
     *
     * @see {@link ui/ComponentOverride}
     *
     * @type {Function|Element}
     * @public
     */
    tooltipComponent: _propTypes["default"].componentOverride,

    /**
     * The value of the slider.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    value: _propTypes2["default"].number
  },
  render: function render(_ref) {
    var orientation = _ref.orientation,
        proportion = _ref.proportion,
        tooltipComponent = _ref.tooltipComponent,
        value = _ref.value,
        rest = _objectWithoutProperties(_ref, _excluded);

    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", _objectSpread(_objectSpread({}, rest), {}, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ComponentOverride["default"], {
        component: tooltipComponent,
        orientation: orientation,
        proportion: proportion,
        children: value
      })
    }));
  }
});
exports.Knob = Knob;
var _default = Knob;
exports["default"] = _default;
