// @enact/sandstone/VideoPlayer/Feedback.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.FeedbackBase = exports.Feedback = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _util = require("../internal/util");

var _FeedbackIcon = _interopRequireDefault(require("./FeedbackIcon"));

var _FeedbackIcons = _interopRequireDefault(require("./FeedbackIcons.js"));

var _FeedbackModule = _interopRequireDefault(require("./Feedback.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["children", "playbackState"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * Feedback {@link sandstone/VideoPlayer}. This displays the media's playback rate and other
 * information.
 *
 * @class Feedback
 * @memberof sandstone/VideoPlayer
 * @ui
 * @private
 */
var FeedbackBase = (0, _kind["default"])({
  name: 'Feedback',
  propTypes:
  /** @lends sandstone/VideoPlayer.Feedback.prototype */
  {
    children: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].number]),

    /**
     * Refers to one of the following possible media playback states.
     * `'play'`, `'pause'`, `'rewind'`, `'fastForward'` ,
     * `'jumpBackward'`, `'jumpForward'`, `'jumpToStart'`, `'jumpToEnd'`, `'stop'`.
     *
     * Each state understands where its related icon should be positioned, and whether it should
     * respond to changes to the `visible` property.
     *
     * This string feeds directly into {@link sandstone/FeedbackIcon.FeedbackIcon}.
     *
     * @type {('play'|'pause'|'rewind'|'fastForward'|'jumpBackward'|'jumpForward'|'jumpToStart'|'jumpToEnd'|'stop')}
     * @public
     */
    playbackState: _propTypes["default"].oneOf(Object.keys(_FeedbackIcons["default"])),

    /**
     * If the current `playbackState` allows this component's visibility to be changed,
     * this component will be hidden. If not, setting this property will have no effect.
     * All `playbackState`s respond to this property except the following:
     * `'rewind'`, `'fastForward'`.
     *
     * @type {Boolean}
     * @default true
     * @public
     */
    visible: _propTypes["default"].bool
  },
  defaultProps: {
    visible: true
  },
  styles: {
    css: _FeedbackModule["default"],
    className: 'feedback'
  },
  computed: {
    className: function className(_ref) {
      var styler = _ref.styler,
          visible = _ref.visible;
      return styler.append({
        hidden: !visible
      });
    },
    children: function children(_ref2) {
      var _children = _ref2.children,
          s = _ref2.playbackState;

      if (_FeedbackIcons["default"][s]) {
        // Working with a known state, treat `children` as playbackRate
        if (_FeedbackIcons["default"][s].message && _children !== 1) {
          // `1` represents a playback rate of 1:1
          return _children.toString().replace(/^-/, '') + _FeedbackIcons["default"][s].message;
        }
      } else {
        // Custom Message
        return _children;
      }
    }
  },
  render: function render(_ref3) {
    var children = _ref3.children,
        playbackState = _ref3.playbackState,
        rest = _objectWithoutProperties(_ref3, _excluded);

    delete rest.visible;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", _objectSpread(_objectSpread({}, rest), {}, {
      children: [_FeedbackIcons["default"][playbackState] && _FeedbackIcons["default"][playbackState].position === 'before' ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_FeedbackIcon["default"], {
        children: playbackState
      }) : null, children ? /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: _FeedbackModule["default"].message,
        children: children
      }) : null, _FeedbackIcons["default"][playbackState] && _FeedbackIcons["default"][playbackState].position === 'after' ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_FeedbackIcon["default"], {
        children: playbackState
      }) : null]
    }));
  }
});
exports.FeedbackBase = FeedbackBase;
var Feedback = (0, _util.onlyUpdateForProps)(FeedbackBase, ['children', 'playbackState', 'visible']);
exports.Feedback = Feedback;
var _default = Feedback;
exports["default"] = _default;
